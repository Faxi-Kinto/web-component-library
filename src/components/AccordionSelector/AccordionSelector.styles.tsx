import styled, { css } from 'styled-components';
import { theme } from '../../config';
import { pxToRem } from '../../styles/basics';
import { phablet } from '../../styles/breakpoints';
import { padding } from '../../styles/basics';

export const Container = styled.div`
  width: ${pxToRem('430px')};
  background-color: ${theme.pallet.WHITE};
  border: 1px solid ${theme.pallet.DISABLED};

  > {
    &:nth-child(n + 2) {
      border-top: 1px solid ${theme.pallet.DISABLED};
    }
  }
  .option {
    cursor: pointer;
    ${padding(pxToRem('20px'), pxToRem('40px'), pxToRem('35px'))};
    min-height: ${pxToRem('100px')};
    &__radio-button {
      border: none;
      align-items: flex-start;
      height: ${pxToRem('40px')};
      span {
        color: ${theme.pallet.LIGHT_BLUE_2};
        /* TODO: remove font-weight when font is ready */
        font-weight: 500;
      }
      svg {
        margin-right: ${pxToRem('12px')};
        margin-left: 0;
      }
    }

    &__description {
      display: block;
      padding-left: 28px;
    }

    &--selected {
      .option__child-container {
        max-height: 100rem;
        transition: max-height 0.3s ease-in-out;
        > * {
          opacity: 1;
          transition: opacity 0.2s ease-in-out 0.3s;
          &:first-child {
            margin-top: 25px;
          }
        }
      }
    }

    &--not-selected {
      .option__radio-button {
        span {
          /* TODO: remove font-weight when font is ready */
          font-weight: 200;
        }
      }
      .option__description {
        color: ${theme.pallet.DISABLED};
      }
      .option__child-container {
        max-height: 0;
        overflow: hidden;
        transition: max-height 0.1s linear;
        > * {
          opacity: 0;
          transition-property: opacity;
        }
      }
    }
  }
  ${phablet(css`
    width: 100%;
    border: none;
    > {
    &:nth-child(n + 1) {
      border-top: 1px solid ${theme.pallet.DISABLED};
    }
    .option {
      ${padding(pxToRem('25px'), '0')};
      > * {
        padding-left:${pxToRem('30px')};
        padding-right:${pxToRem('40px')};
      }
    }
  `)};
`;
