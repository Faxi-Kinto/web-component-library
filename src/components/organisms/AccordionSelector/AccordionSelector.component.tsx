import React, { useState, createRef, useMemo } from 'react';
import * as Styled from './AccordionSelector.styles';
import RadioField from 'components/molecules/RadioField';
import Text from 'components/atoms/Text';
import { FieldProps } from 'components/molecules/FormField/FieldProps';

/**
 * @name AccordionSelector
 *
 * [Insert omponent description]
 *
 * @returns {JSX}
 */

export type AccordionSelectorProps = FieldProps & {
  options: AccordionOption[];
  name: string;
  value?: string;
  defaultSelected?: string;
  className?: string;
};

export type AccordionOption = {
  label?: string;
  child?: React.ReactNode;
  text?: string;
};

const AccordionSelector: React.FC<AccordionSelectorProps> = (
  props: AccordionSelectorProps
): JSX.Element => {
  const { options, name, value, onChange, className, defaultSelected } = props;

  const [currentSelected, setCurrentSelected] = useState(defaultSelected);

  const current = useMemo(() => value || currentSelected, [
    currentSelected,
    value,
  ]);

  const refs: React.RefObject<HTMLInputElement>[] = useMemo(
    () => Array.from({ length: options.length }).map(() => createRef()),
    [options.length]
  );
  return (
    <Styled.Container className={className}>
      {options &&
        options.map((option, index) => (
          <div
            key={index}
            className={`option${
              current !== `${index}`
                ? ' option--not-selected'
                : ' option--selected'
            }`}
            onClick={() => {
              refs[index].current?.click();
            }}
          >
            <RadioField
              inputRef={refs[index]}
              value={`${index}`}
              name={name}
              size="md"
              onChange={event => {
                setCurrentSelected(`${index}`);
                onChange && onChange(event.target.value);
              }}
              checked={current === `${index}`}
              label={option.label}
              className="option__radio-button"
            />
            <Text.Body className="option__description">{option.text}</Text.Body>
            {current === `${index}` && (
              <div className="option__child-container">{option.child}</div>
            )}
          </div>
        ))}
    </Styled.Container>
  );
};

AccordionSelector.defaultProps = {
  defaultSelected: '0',
} as Partial<AccordionSelectorProps>;

export default AccordionSelector;
