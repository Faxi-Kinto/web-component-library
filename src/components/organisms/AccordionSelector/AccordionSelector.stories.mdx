import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";

import AccordionSelector, { AccordionOption } from "./";
import DropdownList from "../../molecules/DropdownList/DropdownList.component.tsx";
import Input from "../../atoms/Input/Input.component.tsx";
import Text from "../../atoms/Text/Text.component.tsx";

export const listItems = [
  {value: 'test1', label: 'Test value 1'},
  {value: 'test2', label: 'Test value 2'},
  {value: 'test3', label: 'Test value 3'},
  {value: 'test4', label: 'Test value 4'},
  {value: 'test5', label: 'Test value 5'}
];

<Meta title="Core|Organisms/AccordionSelector" />

# AccordionSelector Component

## Implementation

#### Default
<Preview>
  <Story name="default">
    <div style={{ width: '450px' }}>
        <AccordionSelector
          name={'test'}
          options={[
            {
              label: 'Administrator only',
              text:
                'Choose this option if you don’t intend to carpool yourself. Don’t worry, you can change your mind later.',
            },
            {
              label: 'Administrator and Carpooler',
              child: (
                <div className="test">
                  <Text.Body>Country</Text.Body>
                  <DropdownList list={listItems} />
                  <Text.Body>Postcode</Text.Body>
                  <Input error={false}></Input>
                </div>
              ),
              text: 'Choose this option if you intend to carpool as well.',
            },
          ]}
        />
      </div>
  </Story>
</Preview>

## Props

<Props of={AccordionSelector} />


## How to use
- Options is array of AccordionOption objects

```js
export type AccordionOption =
  label?: string;
  child?: React.ReactNode;
  text?: string;
;
```
- For each option you can specify
  - label - text for radio button
  - child - components which renders if radio is selected
  - text - text bellow radio button