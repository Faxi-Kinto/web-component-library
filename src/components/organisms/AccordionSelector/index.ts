import AccordionSelector from './AccordionSelector.component';

export default AccordionSelector;
