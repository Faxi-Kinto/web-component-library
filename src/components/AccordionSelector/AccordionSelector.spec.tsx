import React from 'react';
import renderer from 'react-test-renderer';

import AccordionSelector from './AccordionSelector.component';

test('AccordionSelector renders correctly', () => {
  const component = renderer.create(
    <AccordionSelector
      name={'test'}
      options={[
        {
          label: 'Admin',
          child: (
            <div>
              Admin <button>Ok</button>
            </div>
          ),
          text:
            'Choose this option if you don’t intend to carpool yourself. Don’t worry, you can change your mind later.',
        },
        {
          label: 'Carpooler',
          child: <div>Carpooler</div>,
          text: 'Choose this option if you intend to carpool as well.',
        },
      ]}
    />
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
