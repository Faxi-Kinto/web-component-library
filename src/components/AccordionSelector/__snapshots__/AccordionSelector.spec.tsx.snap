// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionSelector renders correctly 1`] = `
.c0 {
  width: 26.875rem;
  background-color: #FFF;
  border: 1px solid #95A1A6;
}

.c0 >:nth-child(n + 2) {
  border-top: 1px solid #95A1A6;
}

.c0 .option {
  cursor: pointer;
  padding: 1.25rem 2.5rem 2.1875rem 2.5rem;
  min-height: 6.25rem;
}

.c0 .option__radio-button {
  border: none;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  height: 2.5rem;
}

.c0 .option__radio-button span {
  color: #00a5b5;
  font-weight: 500;
}

.c0 .option__radio-button svg {
  margin-right: 0.75rem;
  margin-left: 0;
}

.c0 .option__description {
  display: block;
  padding-left: 28px;
}

.c0 .option--selected .option__child-container {
  max-height: 100rem;
  -webkit-transition: max-height 0.3s ease-in-out;
  transition: max-height 0.3s ease-in-out;
}

.c0 .option--selected .option__child-container > * {
  opacity: 1;
  -webkit-transition: opacity 0.2s ease-in-out 0.3s;
  transition: opacity 0.2s ease-in-out 0.3s;
}

.c0 .option--selected .option__child-container > *:first-child {
  margin-top: 25px;
}

.c0 .option--not-selected .option__radio-button span {
  font-weight: 200;
}

.c0 .option--not-selected .option__description {
  color: #95A1A6;
}

.c0 .option--not-selected .option__child-container {
  max-height: 0;
  overflow: hidden;
  -webkit-transition: max-height 0.1s linear;
  transition: max-height 0.1s linear;
}

.c0 .option--not-selected .option__child-container > * {
  opacity: 0;
  -webkit-transition-property: opacity;
  transition-property: opacity;
}

.c1 {
  position: relative;
  background: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1rem;
  height: 2.8125rem;
  width: 100%;
  border: 0.03125rem solid #95A1A6;
  box-sizing: border-box;
  cursor: pointer;
}

.c1 input {
  display: none;
}

.c1 input:checked + svg .inner {
  display: block;
}

.c1 input:checked + svg + span {
  font-weight: 600;
}

.c1 > svg {
  color: #81c8d1;
  margin: 0 1rem;
}

.c1 > svg.lg {
  width: 1.5625rem;
  height: 1.5625rem;
}

.c1 > svg.md {
  width: 1rem;
  height: 1rem;
}

.c1 > svg .inner {
  display: none;
  -webkit-animation-name: animation-inner;
  animation-name: animation-inner;
  -webkit-animation-duration: 0.35s;
  animation-duration: 0.35s;
}

.c2 {
  padding: 0;
  margin: 0;
  font-size: 1rem;
  color: #4E606A;
  font-weight: 300;
  white-space: pre-wrap;
}

@media (max-width:767px) {
  .c0 {
    width: 100%;
    border: none;
  }

  .c0 >:nth-child(n + 1) {
    border-top: 1px solid #95A1A6;
  }

  .c0 > .option {
    padding: 1.5625rem 0 1.5625rem 0;
  }

  .c0 > .option > * {
    padding-left: 1.875rem;
    padding-right: 2.5rem;
  }
}

<div
  className="c0"
>
  <div
    className="option option--selected"
    onClick={[Function]}
  >
    <label
      className="c1 option__radio-button"
    >
      <input
        checked={true}
        className="radio-input"
        disabled={false}
        name="test"
        onChange={[Function]}
        type="radio"
        value="0"
      />
      <svg
        className="md"
        height="25"
        viewBox="0 0 25 25"
        width="25"
      >
        <circle
          cx="12.5"
          cy="12.5"
          fill="none"
          r="11.5"
          stroke="currentColor"
          strokeWidth="1"
        />
        <circle
          className="inner"
          cx="12.5"
          cy="12.5"
          fill="currentColor"
          r="30%"
        />
      </svg>
      <span
        className="c2"
        color="#4E606A"
        fontSize="16px"
      >
        Admin
      </span>
    </label>
    <span
      className="c2 option__description"
      color="#4E606A"
      fontSize="16px"
    >
      Choose this option if you don’t intend to carpool yourself. Don’t worry, you can change your mind later.
    </span>
    <div
      className="option__child-container"
    >
      <div>
        Admin 
        <button>
          Ok
        </button>
      </div>
    </div>
  </div>
  <div
    className="option option--not-selected"
    onClick={[Function]}
  >
    <label
      className="c1 option__radio-button"
    >
      <input
        checked={false}
        className="radio-input"
        disabled={false}
        name="test"
        onChange={[Function]}
        type="radio"
        value="1"
      />
      <svg
        className="md"
        height="25"
        viewBox="0 0 25 25"
        width="25"
      >
        <circle
          cx="12.5"
          cy="12.5"
          fill="none"
          r="11.5"
          stroke="currentColor"
          strokeWidth="1"
        />
        <circle
          className="inner"
          cx="12.5"
          cy="12.5"
          fill="currentColor"
          r="30%"
        />
      </svg>
      <span
        className="c2"
        color="#4E606A"
        fontSize="16px"
      >
        Carpooler
      </span>
    </label>
    <span
      className="c2 option__description"
      color="#4E606A"
      fontSize="16px"
    >
      Choose this option if you intend to carpool as well.
    </span>
  </div>
</div>
`;
